packages/ui/src/components/Visualization/Custom/hooks/paste-step.hook.tsx

```diff
@@ -3,54 +3,77 @@ import { AddStepMode, IVisualizationNode } from '../../../../models/visualizatio
-          const filter = entitiesContext.camelResource.getCompatibleComponents(
-            mode,
-            vizNode.data,
-            vizNode.getComponentSchema()?.definition,
-          );
+      const pastedNodeType = pastedNodeValue.type;
+      const baseNodeType = entitiesContext.camelResource.getType();
+      const isSameType = pastedNodeType === baseNodeType;
 
-          /** Check paste compatibility */
-          const compatibility = catalogModalContext?.checkCompatibility(pastedNodeValue.name, filter) ?? false;
+      /** Validate the pasted node */
+      if (!isSameType) return false;
+      /** Get compatible nodes and the location where can be introduced */
+      const filter = entitiesContext.camelResource.getCompatibleComponents(
+        mode,
+        vizNode.data,
+        vizNode.getComponentSchema()?.definition,
+      );
 
-          setIsCompatible(compatibility);
-          return;
-        }
-      }
+      /** Check paste compatibility */
+      return catalogModalContext?.checkCompatibility(pastedNodeValue.name, filter) ?? false;
+    },
+    [catalogModalContext, entitiesContext, mode, vizNode],
+  );
 
-      setIsCompatible(false);
+  /** Compatibility check on effect */
+  useEffect(() => {
+    const validate = async () => {
+      try {
+        await navigator.permissions.query({ name: 'clipboard-read' as PermissionName });
+        const pastedNodeValue = await ClipboardManager.paste();
+        const compatible = checkClipboardCompatibility(pastedNodeValue);
+        setIsCompatible(compatible);
+      } catch (error) {
+        // fallback to allow pasting incase of permission issues (for Firefox or other browsers)
+        setIsCompatible(true);
+        console.error('Clipboard read permission not supported:', error);
+      }
     };
 
-    checkCompatibility();
-  }, [entitiesContext, catalogModalContext, mode, vizNode]);
+    validate();
+  }, [checkClipboardCompatibility]);
 
   const onPasteStep = useCallback(async () => {
-    if (!vizNode || !entitiesContext) return;
-
     const pastedNodeValue = await ClipboardManager.paste();
-    if (pastedNodeValue) {
-      /** Paste copied node to the entities */
-      vizNode.pasteBaseEntityStep(pastedNodeValue, mode);
-      /** Update entity */
-      entitiesContext.updateEntitiesFromCamelResource();
+    if (!vizNode || !entitiesContext || !pastedNodeValue) return;
+
+    const compatible = checkClipboardCompatibility(pastedNodeValue);
+    if (!compatible) {
+      /** Open the modal with the invalid paste action information  */
+      await pasteModalContext?.actionConfirmation({
+        title: 'Invalid Paste Action',
+        text: 'Pasted node is not compatible with the current context.',
+        buttonOptions: {},
+      });
+      return;
     }
-  }, [entitiesContext, mode, vizNode]);
+
+    /** Paste copied node to the entities */
+    vizNode.pasteBaseEntityStep(pastedNodeValue, mode);
```

Comment:
We are missing an assertion that this is not called when the pasted node it's not compatible.

Perhaps something like:
```
it('should not call pasteBaseEntityStep() when the step it's not compatible', async () => {
```

Reply:
Added the test for this scenario.